Add PoweredHingeJoint.proto.

brettle opened the pull request:
Also:
* Change DBSExample.wbt to use PoweredHingeJoint.
* Reorganize the Encoder proto hierarchy to distinguish builtin encoders from standalone encoders.
* Fix bug in SimRegisterer caused by the non-uniqueness of encoder names.

brettle commented:
Fixes #51.

CoolSpy3 commented:
About to take a look at this, but I realized the windows CI has been running for almost 3 hours. Any ideas on why that would happen? It works for me locally.

brettle commented:
> About to take a look at this, but I realized the windows CI has been running for almost 3 hours. Any ideas on why that would happen? It works for me locally.

Not off hand. Looking into it. Could it be related to your earlier CI changes?

Not sure if [this](https://github.com/DeepBlueRobotics/DeepBlueSim/actions/runs/9308041120/job/25620581468?pr=71#step:6:256) is normal.

CoolSpy3 commented:
> Not off hand. Looking into it. Could it be related to your earlier CI changes?
That was my first thought as well, but I'm not 100% sure how.
> Not sure if [this](https://github.com/DeepBlueRobotics/DeepBlueSim/actions/runs/9308041120/job/25620581468?pr=71#step:6:256) is normal.
I think [so](https://github.com/DeepBlueRobotics/DeepBlueSim/actions/runs/9299643395/job/25593981351#step:6:1273).

CoolSpy3 commented:
Looking at it, it doesn't appear that the `check` task ever actually started. (ie. the program's still waiting for it to start up) :/

brettle commented:
> Looking at it, it doesn't appear that the `check` task ever actually started. (ie. the program's still waiting for it to start up) :/

The `systemTest` task [is running](https://github.com/DeepBlueRobotics/DeepBlueSim/actions/runs/9308041120/job/25620581468?pr=71#step:6:879) so `check` must be running. I've downloaded a copy of the log. I'm going to 1) update the branch, and 2) add a `Timeout` annotation to the `SystemTestRobot` class. That should at least give us access to the webots log which might provide more clues.

CoolSpy3 reviewed (changes requested):
There's still something weird going on here with the encoder base protos. (Note, when I refer to deleting parameters, I am not referring to `noiseStdDevRad`. That should remain exposed.)

`WPIEncoderBase` expects a name field and forwards it to `PositionSensor`, but it still accepts and discards the other fields.

`WPIBuiltinEncoderBase` passes in all of the old parameters (which are unused), and leaves the name as `""`.

`WPIStandaloneEncoderBase` manually calculates the name, but still passes in the other values.

---

What I *think* makes sense, based on the other changes, is to remove the now unused fields from `WPIEncoderBase`. (This makes it the same as `PositionSensor`, so we could also remove the proto entirely and just use `PositionSensor` directly.)

`WPIBuiltinEncoderBase` should either be exposed directly (and `REVBuiltinEncoder` deleted) or should accept some parameter indicating the *type* of builtin encoder. (Otherwise, there's no distinction between it and its subtypes.) It should then be updated to have a name of something like `DBSim_BuiltinEncoder` or `DBSim_BuiltinEncoder_Type` to remain consistent with our other protos. As stated in the comment, the controller will determine the other necessary parameters. They are unneeded and can be deleted.

`WPIStandalineEncoderBase` already does the name calculation, so there's no need to pass the other parameters to it's super-object.

brettle commented:
I can't reproduce it.

1. [The ci run that started when I updated the branch](https://github.com/DeepBlueRobotics/DeepBlueSim/actions/runs/9310163935) succeeded on everything.
2. [The ci run that started when I added the timeout](https://github.com/DeepBlueRobotics/DeepBlueSim/actions/runs/9310198282) also succeeded on Windows but [timed out out on macos only seconds before the test would have finished (!)](https://github.com/DeepBlueRobotics/DeepBlueSim/actions/runs/9310198282/job/25627135723#step:6:1671).
3. [The ci run that hung and I restarted](https://github.com/DeepBlueRobotics/DeepBlueSim/actions/runs/9308041120) succeeded on everything.
4. The webots log always appears to be empty.

I'm going to bump the timeout up to 25 minutes in hopes of at least making the test pass on macos consistently.

brettle commented:
> I'm going to bump the timeout up to 25 minutes in hopes of at least making the test pass on macos consistently.

You beat me to it.

brettle commented:
> There's still something weird going on here with the encoder base protos.

To help you understand where I was coming from... I had *thought* that in `PoweredHingeJoint.proto` if I said:

```
    field MFNode{WPIMotorBase{}, WPIEncoderBase{}, Brake{}} devices [NEOMotor{} REVBuiltinEncoder{} Brake{}]
```
then any current or future protos that derive from `WPIMotorBase` or `WPIEncoderBase` would be allowed. I also thought I'd be able to tell in the js whether a particular encoder device was a subclass of `WPIBuiltInEncoderBase` so I'd know whether it was prohibited with a PWM controller. Unfortunately, Webots isn't currently smart enough to do either of those things at the moment. But, if it is in the future, the extra layers in the hierarchy will be useful. Since I'd already done the work, I figure I might as well leave them in.

In terms of how the actual encoder protos behave, I was trying to minimize changes to the actual java code.

> What I _think_ makes sense, based on the other changes, is to remove the now unused fields from `WPIEncoderBase`. 

Done.

> (This makes it the same as `PositionSensor`, so we could also remove the proto entirely and just use `PositionSensor` directly.)

I'd rather keep it as a marker interface, a place for potentially adding fields that are frc specific across all encoders, and a way of possibly easily limiting which devices are allowed in other protos if Webots gets that capability.

> `WPIBuiltinEncoderBase` should either be exposed directly (and `REVBuiltinEncoder` deleted) or should accept some parameter indicating the _type_ of builtin encoder. (Otherwise, there's no distinction between it and its subtypes.)

The type of the builtin encoder is the name of the proto. Right now there is only one subclass and it doesn't need any extra fields because the motor and its controller fully determine its details. Other subclasses might benefit from those fields.

> It should then be updated to have a name of something like `DBSim_BuiltinEncoder` or `DBSim_BuiltinEncoder_Type` to remain consistent with our other protos.

I didn't do the name mangling because I didn't want to trigger [the current java code](https://github.com/DeepBlueRobotics/DeepBlueSim/blob/af561faf461992a677d3ba7112865bfdd1039be3/plugin/controller/src/main/java/org/team199/deepbluesim/SimRegisterer.java#L46) to create an additional encoder mediator when [it already creates one](https://github.com/DeepBlueRobotics/DeepBlueSim/blob/af561faf461992a677d3ba7112865bfdd1039be3/plugin/controller/src/main/java/org/team199/deepbluesim/mediators/CANMotorMediator.java#L50) when it connects the motor. If that logic doesn't work for a future built-in encoder, it's proto can do the name mangling.

> As stated in the comment, the controller will determine the other necessary parameters. They are unneeded and can be deleted.

Per above, in the future there might be additional subclasses that might need some of the other fields (e.g. the TalonFX encoder is actually an absolute encoder). `WPIBuiltinEncoderBase` enables that.

> `WPIStandalineEncoderBase` already does the name calculation, so there's no need to pass the other parameters to it's super-object.

Fair enough. I've removed them.

brettle commented on plugin/controller/src/webotsFolder/dist/protos/WPIMotorBase.proto:
```diff
@@ -1,5 +1,6 @@
 #VRML_SIM R2023b utf8
 # template language: javascript
+# tags: hidden, no3dView
```
Removed.

CoolSpy3 reviewed (approved):
I made a couple of small changes to `WPIBuiltinEncoderBase.proto`, but everything else looks good!
